isign/bundle.py

```diff
@@ -0,0 +1,159 @@
+        'iPhoneOS' in plist['CFBundleSupportedPlatforms']
+    )
+
+
+class Bundle(object):
+    """ A bundle is a standard directory structure, a signable, installable set of files.
+        Apps are Bundles, but so are some kinds of Frameworks (libraries) """
+    helpers = []
+    executable_class = None
+
+    def __init__(self, path):
+        self.path = path
+        is_native = self.precheck()
+        if not is_native:
+            raise NotMatched("not a native iOS bundle")
+
+        # will be added later
+        self.seal_path = None
+
+        info_path = join(self.path, 'Info.plist')
+        self.info = biplist.readPlist(info_path)
+
+    def precheck(self):
+        """ Checks if a path looks like this kind of app,
+            return stuff we'll need to know about its structure """
+        is_native = False
+        plist_path = join(self.path, "Info.plist")
+        if exists(plist_path):
+            log.debug("got a plist path, {}".format(plist_path))
+            plist = biplist.readPlist(plist_path)
+            is_native = is_info_plist_native(plist)
+            log.debug("is native: {}".format(is_native))
+        return is_native
+
+    def get_executable_path(self):
+        """ Path to the main executable. For an app, this is app itself. For
+            a Framework, this is the main framework """
+        executable_name = None
+        if 'CFBundleExecutable' in self.info:
+            executable_name = self.info['CFBundleExecutable']
+        else:
+            executable_name, _ = splitext(basename(self.path))
+        executable = join(self.path, executable_name)
+        if not exists(executable):
+            raise Exception(
+                'could not find executable for {0}'.format(self.path))
+        return executable
+
+    def sign(self, signer):
+        """ Sign everything in this bundle, recursively with sub-bundles """
+        # log.debug("SIGNING: %s" % self.path)
+        frameworks_path = join(self.path, 'Frameworks')
+        if exists(frameworks_path):
+            # log.debug("SIGNING FRAMEWORKS: %s" % frameworks_path)
+            # sign all the frameworks
+            for framework_name in os.listdir(frameworks_path):
+                framework_path = join(frameworks_path, framework_name)
+                # log.debug("checking for framework: %s" % framework_path)
+                try:
+                    framework = Framework(framework_path)
+                except NotMatched:
+                    # log.debug("not a framework: %s" % framework_path)
+                    continue
+                # log.debug("resigning: %s" % framework_path)
+                framework.resign(signer)
+            # sign all the dylibs
+            dylib_paths = glob.glob(join(frameworks_path, '*.dylib'))
+            for dylib_path in dylib_paths:
+                dylib = signable.Dylib(dylib_path)
+                dylib.sign(self, signer)
+        # then create the seal
```

Comment:
http://cache1.asset-cache.net/gc/486189893-recording-artist-seal-attends-the-8th-annual-gettyimages.jpg?v=1&c=IWSAsset&k=2&d=X7WJLa88Cweo9HktRLaNXkpj%2FCW2j63gMuiLn%2BTSjcfcKhmZFad32p0ArpgRd7RNVjGCSI8DwO%2ByPB4DKW8W8w%3D%3D